BMS_Software.ino

```diff
@@ -0,0 +1,1155 @@
+static uint8_t cell_undervoltage_count = 0;
+static bool low_voltage_state = false;
+static int num_low_voltage_reminder = 0;
+static int time_of_low_voltage = 0;
+
+// Temp
+static int num_meas_batt_temp = 0;
+static bool batt_temp_avail = false;
+static bool overtemp_state = false;
+static bool fans_on = false;
+
+// Logic Switch
+static bool forgotten_logic_switch = false;
+static int num_out_voltage_loops = 0;
+static int time_switch_forgotten = 0;
+static int time_switch_reminder = 0;
+static bool estop_released_beep = false;
+
+// Functions /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void setInputPins()
+{
+  pinMode(PACK_I_MEAS_PIN, INPUT); // pack current sensor
+  pinMode(PACK_V_MEAS_PIN, INPUT); // pack voltage sensor
+  pinMode(TEMP_degC_MEAS_PIN, INPUT); // temperature sensor
+  pinMode(C1_V_MEAS_PIN, INPUT);      //cell voltage sensor
+  pinMode(C2_V_MEAS_PIN, INPUT);
+  pinMode(C3_V_MEAS_PIN, INPUT);
+  pinMode(C4_V_MEAS_PIN, INPUT);
+  pinMode(C5_V_MEAS_PIN, INPUT);
+  pinMode(C6_V_MEAS_PIN, INPUT);
+  pinMode(C7_V_MEAS_PIN, INPUT);
+  pinMode(C8_V_MEAS_PIN, INPUT);
+
+  return;
+} // end func
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void setOutputPins() // output pin functions
+{
+  pinMode(BUZZER_CTR_PIN, OUTPUT);  //buzzer control pin
+  pinMode(FAN_CTR_PIN, OUTPUT); //all-fan control  pin
+  pinMode(SER_TX_IND, OUTPUT);  //LCD communication pin
+  pinMode(SW_IND_PIN, OUTPUT); //software indicator pin
+  pinMode(SW_ERR_PIN, OUTPUT);  //software error pin
+
+  return;
+} // end func
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void setOutputStates()
+{
+  digitalWrite(BUZZER_CTR_PIN, LOW);  //turn off buzzer
+  digitalWrite(FAN_CTR_PIN, LOW); //turn off fan
+  digitalWrite(SER_TX_IND, LOW);  //turn off LCD communication
+  digitalWrite(SW_IND_PIN, LOW);  //turn off software indicator LED
+  digitalWrite(SW_ERR_PIN, LOW);  //turn off software error LED
+
+  return;
+} // end func
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void getMainCurrent(float &main_current)
+{
+  ////Serial.print("adc current:  ");
+  ////Serial.println(analogRead(PACK_I_MEAS_PIN));
+
+  main_current =  map(analogRead(PACK_I_MEAS_PIN), CURRENT_ADC_MIN, CURRENT_ADC_MAX, CURRENT_MIN, CURRENT_MAX); //fetch pack current
+  // display of current functions
+  ////Serial.print("current:  ");
+  ////Serial.println(main_current);
+
+  if (main_current > OVERCURRENT)		//check current > overcurrent, if not return
+  {
+    delay(DEBOUNCE_DELAY); // debounce delay is to check twice in a short period of time
```

Comment:
There might be a Teensy Library for debouncing to correctly debounce the reading